modules/compiler/vecz/test/lit/llvm/scalarize-bitcast.ll

```diff
@@ -0,0 +1,111 @@
+; CHECK-NEXT: [[B2:%.+]] = lshr i32 [[B1]], 16
+; CHECK-NEXT: [[B3:%.+]] = trunc i32 [[B2]] to i16
+; CHECK-NEXT: [[B4:%.+]] = bitcast i16 [[B3]] to half
+; CHECK-NEXT: [[C1:%.+]] = bitcast float [[C0]] to i32
+; CHECK-NEXT: [[C2:%.+]] = trunc i32 [[C1]] to i16
+; CHECK-NEXT: [[C3:%.+]] = bitcast i16 [[C2]] to half
+; CHECK-NEXT: [[D1:%.+]] = bitcast float [[C0]] to i32
+; CHECK-NEXT: [[D2:%.+]] = lshr i32 [[D1]], 16
+; CHECK-NEXT: [[D3:%.+]] = trunc i32 [[D2]] to i16
+; CHECK-NEXT: [[D4:%.+]] = bitcast i16 [[D3]] to half
+; CHECK:      store half [[A3]],
+; CHECK-NEXT: store half [[B4]],
+; CHECK-NEXT: store half [[C3]],
+; CHECK-NEXT: store half [[D4]],
+; CHECK-NEXT: ret void
+
+define dso_local spir_kernel void @bitcast2(ptr addrspace(1) %in, ptr addrspace(1) %out) {
+entry:
+  %gid = tail call i64 @__mux_get_global_id(i32 noundef 0)
+  %pin = getelementptr inbounds <4 x half>, ptr addrspace(1) %in, i64 %gid
+  %pout = getelementptr inbounds <2 x float>, ptr addrspace(1) %out, i64 %gid
+  %0 = load <4 x half>, ptr addrspace(1) %pin, align 4
+  %1 = bitcast <4 x half> %0 to <2 x float>
+  store <2 x float> %1, ptr addrspace(1) %pout, align 4
+  ret void
+}
+
+; CHECK-LABEL: define{{.*}}spir_kernel void @__vecz_v4_bitcast2
+; CHECK:      [[A0:%.+]] = load half,
+; CHECK-NEXT: [[B0:%.+]] = load half,
+; CHECK-NEXT: [[C0:%.+]] = load half,
+; CHECK-NEXT: [[D0:%.+]] = load half,
+; CHECK-NEXT: [[A1:%.+]] = bitcast half [[A0]] to i16
+; CHECK-NEXT: [[A2:%.+]] = zext i16 [[A1]] to i32
+; CHECK-NEXT: [[B1:%.+]] = bitcast half [[B0]] to i16
+; CHECK-NEXT: [[B2:%.+]] = zext i16 [[B1]] to i32
+; CHECK-NEXT: [[B3:%.+]] = shl i32 [[B2]], 16
+; CHECK-NEXT: [[AB4:%.+]] = or i32 [[A2]], [[B3]]
+; CHECK-NEXT: [[AB5:%.+]] = bitcast i32 [[AB4]] to float
+; CHECK-NEXT: [[C1:%.+]] = bitcast half [[C0]] to i16
+; CHECK-NEXT: [[C2:%.+]] = zext i16 [[C1]] to i32
+; CHECK-NEXT: [[D1:%.+]] = bitcast half [[D0]] to i16
+; CHECK-NEXT: [[D2:%.+]] = zext i16 [[D1]] to i32
+; CHECK-NEXT: [[D3:%.+]] = shl i32 [[D2]], 16
+; CHECK-NEXT: [[CD4:%.+]] = or i32 [[C2]], [[D3]]
+; CHECK-NEXT: [[CD5:%.+]] = bitcast i32 [[CD4]] to float
+; CHECK:      store float [[AB5]],
+; CHECK-NEXT: store float [[CD5]],
+; CHECK-NEXT: ret void
+
+define dso_local spir_kernel void @bitcast3(ptr addrspace(1) %in, ptr addrspace(1) %out) {
+entry:
+  %gid = tail call i64 @__mux_get_global_id(i32 noundef 0)
+  %pin = getelementptr inbounds <2 x i32>, ptr addrspace(1) %in, i64 %gid
+  %pout = getelementptr inbounds <2 x float>, ptr addrspace(1) %out, i64 %gid
+  %0 = load <2 x i32>, ptr addrspace(1) %pin, align 4
+  %1 = bitcast <2 x i32> %0 to <2 x float>
+  store <2 x float> %1, ptr addrspace(1) %pout, align 4
+  ret void
+}
+
+; CHECK-LABEL: define{{.*}}spir_kernel void @__vecz_v4_bitcast3
+; CHECK:      [[A0:%.+]] = load i32,
+; CHECK-NEXT: [[B0:%.+]] = load i32,
+; CHECK-NEXT: [[A1:%.+]] = bitcast i32 [[A0]] to float
+; CHECK-NEXT: [[B1:%.+]] = bitcast i32 [[B0]] to float
```

Comment:
The original example relies on other passes too and does not demonstrate the problem with the `RUN` line we have here, so it would need some tweaking, but it would be tricky. The problem with the original example is that we ended up with `or i128 %2, poison` when any element was `poison`. I do not think I can add a test that if any element is `poison`, we do not produce `poison`, because it is correct that for that element, we do produce `poison` in that case. I can only test that one element does not wrongly depend on another element (in the original example, that we do not use any `i128` at all), by testing that `A1` only depends on `A0`, and `B1` only depends on `B0`. But I'll have an extra think to see if there is a less obvious way that more directly does test `poison` specifically.

Reply:
Managed to add a test that previously specifically failed in the same way as the original test, does that look good?